src/dust/coagulation/coagulation.cpp

```diff
@@ -0,0 +1,308 @@
+        << std::log(mmax / mmin) / (std::log(std::sqrt(2.0))) + 1. << " instead of " << nm
+        << std::endl;
+    PARTHENON_FAIL(msg);
+  }
+
+  ParArray1D<Real> dust_size("dustSize", nm);
+  auto dust_size_host = dust_size.GetHostMirror();
+
+  for (int n = 0; n < nm; ++n) {
+    Real mgrid = std::exp(conc + cond * n);
+    dust_size_host(n) = std::pow(3.0 * mgrid / (4.0 * M_PI * cpars.rho_p), 1. / 3.);
+  }
+  dust_size.DeepCopy(dust_size_host);
+
+  // allocate array and assign values
+  cpars.klf = ParArray2D<int>("klf", nm, nm);
+  cpars.mass_grid = ParArray1D<Real>("mass_grid", nm);
+  const int n2DRv = coag2DRv::last2;
+  cpars.coagR3D = ParArray3D<Real>("coagReal3D", n2DRv, nm, nm);
+  cpars.cpod_notzero = ParArray3D<int>("idx_nzcpod", nm, nm, 4);
+  cpars.cpod_short = ParArray3D<Real>("nzcpod", nm, nm, 4);
+
+  if (Dust::cgsunit == NULL) {
+    Dust::cgsunit = new CGSUnit;
+  }
+  if (!Dust::cgsunit->isSet()) {
+    Dust::cgsunit->SetCGSUnit(pin);
+  }
+  const Real dfloor = pin->GetOrAddReal("dust", "dfloor", 1.0e-25);
+  // convert to CGS unit
+  cpars.dfloor = dfloor * Dust::cgsunit->mass0 / Dust::cgsunit->vol0;
+  Real a = std::log10(mmin / mmax) / static_cast<Real>(1 - nm);
+
+  initializeArray(nm, cpars.pGrid, cpars.rho_p, cpars.chi, a, dust_size, cpars.klf,
+                  cpars.mass_grid, cpars.coagR3D, cpars.cpod_notzero, cpars.cpod_short);
+
+  params.Add("coag_pars", cpars);
+
+  // other parameters for coagulation
+  const int nstep1Coag = pin->GetOrAddReal("problem", "nstep1Coag", 50);
+  params.Add("nstep1Coag", nstep1Coag);
+  Real dtCoag = 0.0;
+  params.Add("dtCoag", dtCoag, Params::Mutability::Restart);
+  const Real alpha = pin->GetOrAddReal("dust", "coag_alpha", 1.e-3);
+  params.Add("coag_alpha", alpha);
+  const int scr_level = pin->GetOrAddReal("dust", "coag_scr_level", 0);
+  params.Add("coag_scr_level", scr_level);
+
+  return coag;
+}
+
+// using Kokkos par_for() loop
+void initializeArray(const int nm, int &pGrid, const Real &rho_p, const Real &chi,
+                     const Real &a, const ParArray1D<Real> dsize, ParArray2D<int> klf,
+                     ParArray1D<Real> mass_grid, ParArray3D<Real> coag3D,
+                     ParArray3D<int> cpod_notzero, ParArray3D<Real> cpod_short) {
+
+  int ikdelta = coag2DRv::kdelta;
+  auto kdelta = Kokkos::subview(coag3D, ikdelta, Kokkos::ALL, Kokkos::ALL);
+  int icoef_fett = coag2DRv::coef_fett;
+  auto coef_fett = Kokkos::subview(coag3D, icoef_fett, Kokkos::ALL, Kokkos::ALL);
+  parthenon::par_for(
+      parthenon::loop_pattern_flatrange_tag, "initializeCoag1", parthenon::DevExecSpace(),
+      0, nm - 1, KOKKOS_LAMBDA(const int i) {
+        // initialize in_idx(*) array
+        // initialize kdelta array
+        for (int j = 0; j < nm; j++) {
+          kdelta(i, j) = 0.0;
+        }
+        kdelta(i, i) = 1.0;
+        mass_grid(i) = 4.0 * M_PI / 3.0 * rho_p * std::pow(dsize(i), 3);
```

Comment:
Let's try to avoid `std::pow` for integer powers inside kernels.

Reply:
I can fix that

Reply:
We already have `SQR()` for `std::pow(x,2)`; if `x*x*x` shows up in a lot of places we can add a `CUBE() { return x*x*x; }` function